## Question 003

In [25]:
paragraph = '''As a term, data analytics predominantly refers to an assortment of applications, from basic \nbusiness intelligence (BI), reporting and online analytical processing (OLAP) to various forms of \nadvanced analytics. In that sense, it's similar in nature to business analytics, another umbrella \nterm for approaches to analyzing data -- with the difference that the latter is oriented to \nbusiness uses, while data analytics has a broader focus. The expansive view of the term isn't \nuniversal, though: In some cases, people use data analytics specifically to mean advanced \nanalytics, treating BI as a separate category. Data analytics initiatives can help businesses \nincrease revenues, improve operational efficiency, optimize marketing campaigns and customer \nservice efforts, respond more quickly to emerging market trends and gain a competitive edge \nover rivals -- all with the ultimate goal of boosting business performance. Depending on the \nparticular application, the data that's analyzed can consist of either historical records or new \ninformation that has been processed for real-time analytics uses. In addition, it can come from a \nmix of internal systems and external data sources. At a high level, data analytics methodologies \ninclude exploratory data analysis (EDA), which aims to find patterns and relationships in data, \nand confirmatory data analysis (CDA), which applies statistical techniques to determine whether \nhypotheses about a data set are true or false. EDA is often compared to detective work, while \nCDA is akin to the work of a judge or jury during a court trial -- a distinction first drawn by \nstatistician John W. Tukey in his 1977 book Exploratory Data Analysis. Data analytics can also \nbe separated into quantitative data analysis and qualitative data analysis. The former involves \nanalysis of numerical data with quantifiable variables that can be compared or measured \nstatistically. The qualitative approach is more interpretive -- it focuses on understanding the \ncontent of non-numerical data like text, images, audio and video, including common phrases, \nthemes and points of view.'''
print(paragraph)

As a term, data analytics predominantly refers to an assortment of applications, from basic 
business intelligence (BI), reporting and online analytical processing (OLAP) to various forms of 
advanced analytics. In that sense, it's similar in nature to business analytics, another umbrella 
term for approaches to analyzing data -- with the difference that the latter is oriented to 
business uses, while data analytics has a broader focus. The expansive view of the term isn't 
universal, though: In some cases, people use data analytics specifically to mean advanced 
analytics, treating BI as a separate category. Data analytics initiatives can help businesses 
increase revenues, improve operational efficiency, optimize marketing campaigns and customer 
service efforts, respond more quickly to emerging market trends and gain a competitive edge 
over rivals -- all with the ultimate goal of boosting business performance. Depending on the 
particular application, the data that's analyzed can c

In [28]:
paragraph = paragraph.lower()  # standardize wordings
lines = paragraph.split("\n")  # split paragraph into lines
lines_count = len(lines)
word_to_find = "data"

lines_with_word = sum(1 for line in lines if word_to_find in line)
prob_of_word_in_each_line = lines_with_word / lines_count

print(f"Total number of lines: {lines_count}")
print(f"Number of lines with the word '{word_to_find}': {lines_with_word}")
print(f"Probability of the word '{word_to_find}' occurring in each line: {prob_of_word_in_each_line}")

Total number of lines: 23
Number of lines with the word 'data': 14
Probability of the word 'data' occurring in each line: 0.6086956521739131


## Summary
- The results show that out of the 23 lines in the paragraph, the word 'data' appears in 14 of them
- Therefore, the probability of the word 'data' occurring in each line is approximately 0.6087 or 60.87%

In [32]:
from collections import Counter
paragraph = paragraph.lower()  # standardize wordings
lines = paragraph.split("\n")  # split paragraph into lines
word_counts = Counter()

for line_number, line in enumerate(lines, start=1):
    words = line.split()
    distinct_word_count = len(set(words))
    word_counts[distinct_word_count] += 1

print("Distribution of distinct word counts across all the lines:")
for count, frequency in word_counts.items():
    print(f"Word count: {count}, Frequency: {frequency}")


Distribution of distinct word counts across all the lines:
Word count: 14, Frequency: 8
Word count: 13, Frequency: 6
Word count: 15, Frequency: 1
Word count: 10, Frequency: 1
Word count: 16, Frequency: 1
Word count: 12, Frequency: 3
Word count: 17, Frequency: 1
Word count: 19, Frequency: 1
Word count: 5, Frequency: 1


## Summary
- The distribution results indicate the number of lines that have distinct word counts
- The word count represents the number of distinct words in a line, while the frequency shows how many lines have that specific word count
- For example, there are 8 lines that have 14 distinct words, 6 lines with 13 distinct words and so on

In [41]:
import re
paragraph = paragraph.lower()  # standardize wordings
words = re.findall(r'\b\w+\b', paragraph)  # extract words from the paragraph

word_pairs = zip(words, words[1:])  # create pairs of consecutive words
pair_count = sum(1 for pair in word_pairs if pair == ('data', 'analytics'))
total_count_data = sum(1 for word in words if word == 'data' or word == 'data,')
probability_with_data = pair_count / total_count_data

print(f"Number of word 'data analytics': {pair_count}")
print(f"Number of word 'data': {total_count_data}")
print(f"Probability of the word 'analytics' occurring after the word 'data': {probability_with_data}")


Number of word 'data analytics': 6
Number of word 'data': 18
Probability of the word 'analytics' occurring after the word 'data': 0.3333333333333333


## Summary
- Result shows that ther are 18 occurrences of the word "data" in the paragraph
- The word "analytics" appears immediately after "data" in 6 instances
- Therefore, the probability of the word "analytics" occurring after the word "data" is approximately 0.3333 or 33.33%